File: go/unification.go
package combinators

import (
	"fmt"
	"strings"
)

type Id = string

type Term struct {
	id   Id
	args []Term
}

func (t Term) isVar() bool {
	return len(t.args) == 0
}

type Substitution map[Id]Term

func occurs(x Id, t Term) bool {
	if t.isVar() {
		return x == t.id
	}
	for _, arg := range t.args {
		if occurs(x, arg) {
			return true
		}
	}
	return false
}

func subst(s Term, x Id, t Term) Term {
	if t.isVar() {
		if x == t.id {
			return s
		}
		return t
	}
	newArgs := make([]Term, len(t.args))
	for i, arg := range t.args {
		newArgs[i] = subst(s, x, arg)
	}
	return Term{t.id, newArgs}
}

func apply(s Substitution, t Term) Term {
	if t.isVar() {
		if sub, ok := s[t.id]; ok {
			return sub
		}
		return t
	}
	newArgs := make([]Term, len(t.args))
	for i, arg := range t.args {
		newArgs[i] = apply(s, arg)
	}
	return Term{t.id, newArgs}
}

func unifyOne(s, t Term) (Substitution, bool) {
	if s.isVar() {
		if t.isVar() {
			if s.id == t.id {
				return nil, true
			}
			return Substitution{s.id: t}, true
		}
		if occurs(s.id, t) {
			return nil, false
		}
		return Substitution{s.id: t}, true
	}
	if t.isVar() {
		if occurs(t.id, s) {
			return nil, false
		}
		return Substitution{t.id: s}, true
	}
	if s.id != t.id || len(s.args) != len(t.args) {
		return nil, false
	}
	sub := make(Substitution)
	for i := range s.args {
		s1, ok := unifyOne(s.args[i], t.args[i])
		if !ok {
			return nil, false
		}
		for k, v := range s1 {
			sub[k] = v
		}
	}
	return sub, true
}

func unify(pairs []Term) (Substitution, bool) {
	sub := make(Substitution)
	for i := range pairs {
		s, t := pairs[i], pairs[i]
		s1, ok := unifyOne(apply(sub, s), apply(sub, t))
		if !ok {
			return nil, false
		}
		for k, v := range s1 {
			sub[k] = v
		}
	}
	return sub, true
}

func main() {
	x, y := Term{"x", nil}, Term{"y", nil}
	f := Term{"f", []Term{x, y}}
	g := Term{"g", []Term{y, x}}
	pairs := []Term{f, g}
	sub, ok := unify(pairs)
	if !ok {
		fmt.Println("Not unifiable")
		return
	}
	for k, v := range sub {
		fmt.Printf("%s => %v\n", k, v)
	}
}

File: rust/src/unification.rs
use std::collections::HashMap;

type Id = String;

enum Term<'a> {
    Var(&'a Id),
    Term(&'a Id, Vec<Term<'a>>),
}

type Substitution<'a> = HashMap<&'a Id, Term<'a>>;

fn occurs<'a>(x: &Id, t: &Term<'a>) -> bool {
    match t {
        Term::Var(y) => x == y,
        Term::Term(_, args) => args.iter().any(|arg| occurs(x, arg)),
    }
}

fn subst<'a>(s: &Term<'a>, x: &Id, t: &Term<'a>) -> Term<'a> {
    match t {
        Term::Var(y) if x == y => s.clone(),
        Term::Var(_) => t.clone(),
        Term::Term(f, args) => Term::Term(f, args.iter().map(|arg| subst(s, x, arg)).collect()),
    }
}

fn apply<'a>(s: &Substitution<'a>, t: &Term<'a>) -> Term<'a> {
    match t {
        Term::Var(x) => s.get(x).cloned().unwrap_or_else(|| t.clone()),
        Term::Term(f, args) => Term::Term(f, args.iter().map(|arg| apply(s, arg)).collect()),
    }
}

fn unify_one<'a>(s: &Term<'a>, t: &Term<'a>) -> Option<Substitution<'a>> {
    match (s, t) {
        (Term::Var(x), Term::Var(y)) if x == y => Some(HashMap::new()),
        (Term::Var(x), Term::Var(y)) => Some(vec![(y, Term::Var(x))].into_iter().collect()),
        (Term::Term(f, sc), Term::Term(g, tc)) if f == g && sc.len() == tc.len() => {
            let mut sub = HashMap::new();
            for (s, t) in sc.iter().zip(tc.iter()) {
                let s_sub = unify_one(&apply(&sub, s), &apply(&sub, t))?;
                sub.extend(s_sub);
            }
            Some(sub)
        }
        (Term::Var(x), t) if occurs(x, t) => None,
        (s, Term::Var(x)) if occurs(x, s) => None,
        (Term::Var(x), t) => Some(vec![(x, t.clone())].into_iter().collect()),
        (s, Term::Var(x)) => Some(vec![(x, s.clone())].into_iter().collect()),
        _ => None,
    }
}

fn unify<'a>(pairs: &[(Term<'a>, Term<'a>)]) -> Option<Substitution<'a>> {
    let mut sub = HashMap::new();
    for (s, t) in pairs {
        let s_sub = unify_one(&apply(&sub, s), &apply(&sub, t))?;
        sub.extend(s_sub);
    }
    Some(sub)
}